### DISCUSSION TITLE
getStaticProps not returning JSON (SyntaxError: Unexpected token < in JSON at position 0)

### DISCUSSION THREAD

**Discussion by Kvaibhav01** [vercel/next.js#42989] [Help] [answered]
Hello there!

My `getStaticProps` method can't return a JSON while I fetch data with an external API. I want to generate dynamic routes and for that I have this code under _/pages/news/[slug].js_:

```js
export default function NewsDetail({ data }) {
  return (
    <div>
      <p>Title: {data.title}</p>
    </div>
  );
}

export const getStaticPaths = async () => {
  const res = await fetch('http://localhost:3001/en/');
  const data = await res.json();

  const paths = data.map((currElm) => {
    return {
      params: {
        slug: currElm.slug,
      },
    };
  });

  return {
    paths,
    fallback: false,
  };
};

export const getStaticProps = async (context) => {
  const newsSlug = context.params.slug;

  const res = await fetch(`http://localhost:3001/en/${newsSlug}`);
  const data = await res.json(); <-- ERROR -->

  return {
    props: {
      data,
    },
  };
};
```

I get the error:

> SyntaxError: Unexpected token < in JSON at position 0

That definitely means that I am not getting back JSON object. When I change the error line from `const data = await res.json();` to `const data = await res.text();` and log the `data`, it gets me the error page HTML:

```html
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Error</title>
</head>
<body>
<pre>Cannot GET /en/energy-storage-system-market-demand-is-projected-to-reach-44071-gw-in-2028</pre>
</body>
</html>
```

And I checked manually too that if I go to `http://localhost:3001/en/energy-storage-system-market-demand-is-projected-to-reach-44071-gw-in-2028`, it can't GET the page and shows the same HTML rendered page.

My question is, what am I missing in this? Is there an issue with the backend API code itself where something went missing? Or am I doing data fetching and dynamic page generation wrong?

**Comment by icyJoseph** [comment, marked as answer]
If you open the browser console and do:

```js
fetch("http://localhost:3001/en/energy-storage-system-market-demand-is-projected-to-reach-44071-gw-in-2028")
```

What do you get back? try to `.json` or `.text` the response as well. 

I'd say, that it looks like your `:3001` server is expecting something, perhaps `content-type` header, with `application/json`?  does the route even exist on the `:3001` server?


## Edit

OP's server did not contain any route to handle the request it received. I posted a possible code snippet to help out: https://github.com/vercel/next.js/discussions/42989#discussioncomment-4158401

**Comment by Kvaibhav01** [comment]
Hi @icyJoseph thanks for your reply!

When I do `fetch("http://localhost:3001/en/energy-storage-system-market-demand-is-projected-to-reach-44071-gw-in-2028")` with `.json`, then I get the same error on the dynamic route which is:


> SyntaxError: Unexpected token < in JSON at position 0

And when I do `.text`, the page does render but when I log the `data`, I get that HTML error code back.


> I'd say, that it looks like your :3001 server is expecting something, perhaps content-type header, with application/json?

Is this something we should do inside GSP while making the fetch call? 

> does the route even exist on the :3001 server?

When we go to a page like `http://localhost:3001/en/energy-storage-system-market-demand-is-projected-to-reach-44071-gw-in-2028`, it doesn't exist. It says: "Cannot GET /en/energy-storage-system-market-demand-is-projected-to-reach-44071-gw-in-2028".

**Reply by icyJoseph** [reply]
What is the Server running on 3001?

**Reply by Kvaibhav01** [reply]
It's a separate Express project with code something like this:

```js 
// index.js
const express = require('express');
const fs = require('fs');
const path = require('path');

const port = 3001;
const jsonPath = path.join(__dirname, 'data');

function messageJsonFor(locale) {
  const filepath = path.join(jsonPath, `${locale}.json`);
  const contents = fs.readFileSync(filepath, 'utf8');
  const json = JSON.parse(contents);
  return json;
}

const app = express();

app.get('/', (req, res) => {
  res.redirect('/en');
});

app.get('/en', (req, res) => {
  res.json(messageJsonFor('en'));
});

app.get('/es', (req, res) => {
  res.json(messageJsonFor('es'));
});

app.listen(port, () => {
  console.log(`i18n JSON app listening on port ${port}`);
});
```

Rest, it has a folder at root called "data" with two JSON files inside: _en.json_ and _es.json_ containing the news data like:

```json
[
  {
    "id": 1,
    "slug": "energy-storage-system-market-demand-is-projected-to-reach-44071-gw-in-2028",
    "title": "Energy Storage System Market demand is Projected to Reach 440.71 GW in 2028",
    "description": "Energy Storage Systems market is expected to reach 440.71 GW in 2028, growing at a CAGR of 11.1% during 2022- 2028, reports Stratview Research.",
    "content": "This is my first news content",
    "date": "August 16, 2022"
  },
]
```

**Reply by icyJoseph** [reply]
Right, but why would the server respond to `/en/energy-storage-system-market-demand-is-projected-to-reach-44071-gw-in-2028`

Your snippet has only routes for `/en` and `/es`, you'd need something like:

```js
app.get("/:locale(en|es)/:slug", (req, res) => {
  const all = messageJsonFor(req.params.locale);

  const actualData = all.find((entry) => entry.slug === req.params.slug);

  if (!actualData) {
    return res.status(404).json({ message: "Not Found" });
  }

  return res.json(actualData);
});
```

Also that JSON you posted has a floating  `,` : after the data line `},` should be `}`

Please confirm that it works, so I can update my top level answer.

**Reply by Kvaibhav01** [reply]
Thanks @icyJoseph for your explanation! It works. Actually I didn't write the Express code (I'm not experienced in the backend) so I wasn't aware that we would need a new `app.get()` function.

You can update your answer. Have a good day! 😊

**Reply by icyJoseph** [reply]
I updated it, please do mark as answered :) I linked to my comment from there.
